### PR TITLE
make ptr metadata functions callable from stable const fn

### PR DESCRIPTION
**Pull request by RalfJung**
So far this was done with a bunch of `rustc_allow_const_fn_unstable`. But those should be the exception, not the norm. If we are confident we can expose the ptr metadata APIs *indirectly* in stable const fn, we should just mark them as `rustc_const_stable`. And we better be confident we can do that since it's already been done a while ago. ;)

In particular this marks two intrinsics as const-stable: `aggregate_raw_ptr`, `ptr_metadata`. This should be uncontroversial, they are trivial to implement in the interpreter.
Cc @rust-lang/wg-const-eval @rust-lang/lang

### CODE REVIEW

**Comment by RalfJung** [comment]
FWIW this is "reversible" in the sense that the PR doesn't stabilize anything new. Exposing this intrinsic on stable has already happened.

**Comment by workingjubilee** [comment]
Currently the in-flight PR https://github.com/rust-lang/rust/pull/130403 has a `rustc_allow_const_fn_unstable` for this.

But already in the codebase we have these for `intrinsics::ptr_metadata`:
- https://github.com/rust-lang/rust/blob/42ff2eedb0585e32e3e8da0e83ff82dd49987a2c/library/core/src/slice/mod.rs#L114-L117
- https://github.com/rust-lang/rust/blob/42ff2eedb0585e32e3e8da0e83ff82dd49987a2c/library/core/src/ptr/mut_ptr.rs#L1807-L1808

And these for `intrinsics::aggregate_raw_ptr`:
- https://github.com/rust-lang/rust/blob/42ff2eedb0585e32e3e8da0e83ff82dd49987a2c/library/core/src/ptr/mod.rs#L602-L604
- https://github.com/rust-lang/rust/blob/42ff2eedb0585e32e3e8da0e83ff82dd49987a2c/library/core/src/ptr/mod.rs#L628-L630
- https://github.com/rust-lang/rust/blob/42ff2eedb0585e32e3e8da0e83ff82dd49987a2c/library/core/src/ptr/mod.rs#L952-L954

**Comment by RalfJung** [comment]
Indeed, they show up in the diff for this PR as these lines are being removed here.

**Review by scottmcm — APPROVED** [review]
Making the intrinsics const-stable seems clearly good to me.

The stabilization-track stuff is less obvious to me.  But if you're confident that's ok -- it won't show const-stable in the rustdocs, for example -- then sure, `r=me`.

File: library/core/src/ptr/const_ptr.rs [hunk 1]
```diff
@@ -92,7 +92,7 @@ impl<T: ?Sized> *const T {
     /// }
     /// ```
     #[unstable(feature = "set_ptr_value", issue = "75091")]
-    #[rustc_const_unstable(feature = "set_ptr_value", issue = "75091")]
+    #[rustc_const_stable(feature = "ptr_metadata_const", since = "CURRENT_RUSTC_VERSION")]
```
[the remark this answers is not in the document]
Comment (slanterns) [reply library/core/src/ptr/const_ptr.rs:95]: Another example: https://github.com/rust-lang/rust/pull/128228#discussion_r1694120274. tbh it also feels surprising to me 😅

[the remark this answers is not in the document]
Comment (RalfJung) [reply library/core/src/ptr/const_ptr.rs:95]: Yeah I think we should redo how we do const stability. See https://github.com/rust-lang/rust/issues/129815.

But for now this is the system that we got.

**Comment by RalfJung** [comment]
> The stabilization-track stuff is less obvious to me. But if you're confident that's ok -- it won't show const-stable in the rustdocs, for example -- then sure, r=me.

I added a test so I am confident in the most important part, that this is not callable on stable. You can see [here](https://doc.rust-lang.org/nightly/std/task/struct.ContextBuilder.html) how this looks in rustdoc: it has the "This is a nightly-only experimental API" banner. It also shows "const: 1.82.0", which maybe we should suggest they hide. But IMO it's still pretty clear that this is an unstable function.

@bors r=scottmcm

**Pull request merged**